src/functional/ffront/foast_passes/closure_var_folding.py

```diff
@@ -0,0 +1,46 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2022, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either vervisit_Constantsion 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from typing import Any
+
+import functional.ffront.field_operator_ast as foast
+from eve import NodeTranslator, traits
+from eve.utils import ConstantNamespace
+
+
+class ClosureVarFolding(traits.VisitorWithSymbolTableTrait, NodeTranslator):
+    closure_vars: dict[str, Any]
+
+    def __init__(self, closure_vars: dict[str, Any]):
+        self.closure_vars = closure_vars
+
+    @classmethod
+    def apply(cls, node: foast.FieldOperator, closure_vars: dict[str, Any]):
+        return cls(closure_vars=closure_vars).visit(node)
+
+    def visit_Name(self, node: foast.Name, **kwargs):
+        if node.id in self.closure_vars:
+            value = self.closure_vars[node.id]
+            if isinstance(value, ConstantNamespace):
+                return foast.Constant(value=value, location=node.location)
+        return node
+
+    def visit_Attribute(self, node: foast.Attribute, **kwargs):
+        value = self.visit(node.value)
+        if isinstance(value, foast.Constant):
+            if hasattr(value.value, node.attr):
+                return foast.Constant(value=getattr(value.value, node.attr), location=node.location)
+            # TODO: throw a syntax error with all necessary information.
```

Comment:
I left a todo because this cannot be fixed in this PR. I already told Rico in his previous PR that I think this is a systematic problem that we don't have proper exception classes. I cannot use `FieldOperatorSyntaxError` because it's defined in a parent package. I'm not going to define a `ClosureVarFoldingSyntaxError`, because that's against the proper way of using exceptions - and so is `FieldOperatorSyntaxError` for that matter. Exceptions should be subclassed based on the error that they represent, not based on the location where they are raised.

Fixing this problem would require introducing a `SyntaxError` that's consistently used across the whole of gt4py, and potentially subclassing it with `AttributeNotFoundError` to indicate accessing an inexistent attribute of an object in the AST.

But for this, we have to actually sit down as a team, design an analogue to Python's exception classes or `<stdexcept>`, and supplement the basic exceptions used in the whole project with module-specific ones subclassed from the basic ones.

At least, this is my take on the situation.

Reply:
As discussed.